### DISCUSSION TITLE
Scoping condition on a "local" keyword

### DISCUSSION THREAD

**Discussion by christian-oreilly** [textX/textX#387] [Q&A] [answered]
I am new to textX (and DSL building in general) and I have a hard time getting the scoping right.  I try to salvage a code that has been written by someone else about 5 years ago and which currently crashes. I try to parse a language which has global variables but that can also have local variables using the keyword LOCAL. A minimal example would look something like this:

```python
from textx.metamodel import metamodel_from_str 
from textx.exceptions import TextXError

grammar = """
 Program:
     functions *= Function; 

 Function: 'FUNCTION'
     locals += Locals
     assignements *= Assignment
     'END';

 Local: name=ID;
 Locals: 'LOCAL' vars*=Local[','];
 Var: Local | Variable ; 
 Variable: name=ID;
 Assignment: variable=[Var] '=' value=Number;
 Number: /[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?/;
 """
try:
    meta = metamodel_from_str(grammar, autokwd=True)
    print("Meta-Model OK")
except TextXError as e:
    print("Error in meta-model:")
    print(e)
    
code ='''
FUNCTION
    LOCAL a
    a = 1
    b = 2
END

FUNCTION
    LOCAL a
    a = 1
    b = 2
END

'''
try:
    model = meta.model_from_str(code)
    print("Model ok")
except TextXError as e:
    print("Error in model: ", e)
```

Output:
```
Meta-Model OK
Error in model:  None:4:5: name a is not unique.
```
I know that I need to use `meta.register_scope_providers`. I tried to make it work with `FQN` and `RelativeName` based on https://github.com/textX/textX/issues/107 and https://github.com/textX/textX/issues/170 but I must admit that it is a bit of guesswork and it is not going anywhere. I think I get the main idea of the solutions discussed in these issues, but I have a hard time seeing how I can use the `LOCAL` statement to make the following variables (i.e., `a`) local specifically to the `FUNCTION` block. Any advice would be much appreciated.

**Comment by goto40** [comment]
You could probably solve your problem with rrel: Something Like `Assignment: variable=[Var|ID|..~locals.vars]` etc.. See http://textx.github.io/textX/latest/rrel/

The snippet resolves locals only. For variables you need to extend the rrel expression (add a second path to variables, separated by a comma (see example below).

Note: rrel adds a scope provider for you.

**Comment by goto40** [comment, marked as answer]
I quickly made the demo:
```python
from textx.metamodel import metamodel_from_str 
from textx.exceptions import TextXError

grammar = """
 Program:
     variables *= Variables
     functions *= Function; 

 Function: 'FUNCTION'
     locals += Locals
     assignements *= Assignment
     'END';

 Local: name=ID;
 Locals: 'LOCAL' vars*=Local[','];
 Variables: 'VARIABLES' vars*=Variable[','];
 Var: Local | Variable ; 
 Variable: name=ID;
 Assignment: variable=[Var|ID|..~locals.vars,~variables.vars] '=' value=Number;
 Number: /[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?/;
 """
try:
    meta = metamodel_from_str(grammar, autokwd=True)
    print("Meta-Model OK")
except TextXError as e:
    print("Error in meta-model:")
    print(e)
    
code ='''
VARIABLES b,c

FUNCTION
    LOCAL a
    a = 1
    b = 2
    c = 3
END

FUNCTION
    LOCAL a
    a = 1
    b = 2
END
'''

def run(m):
    for idx,f in enumerate(m.functions):
        print(f"function no {idx}:")
        for a in f.assignements:
            print(f" - assignment: {a.variable.name} ({id(a.variable)}) = {a.value}")

try:
    model = meta.model_from_str(code)#
    print("Model ok")
    run(model)
except TextXError as e:
    print("Error in model: ", e)
```

Output
```
Meta-Model OK
Model ok
function no 0:
 - assignment: a (139871323909344) = 1
 - assignment: b (139871323902288) = 2
 - assignment: c (139871323904976) = 3
function no 1:
 - assignment: a (139871323904016) = 1
 - assignment: b (139871323902288) = 2
```

**Comment by christian-oreilly** [comment]
Thank you very much for taking the time of providing this example. It works fine and it will be useful in teaching myself how this work. However, while trying to apply it to my use cases, I ran into a few more problems. I thought it was better for me make sure that I clearly understand how rrel works. I tried to unpack the lookup tests in here https://github.com/textX/textX/blob/master/tests/functional/test_scoping/test_rrel.py For example, in this case 
https://github.com/textX/textX/blob/22b66e594e60077262c23bcdbe961d827499199a/tests/functional/test_scoping/test_rrel.py#L128

I can relatively easily follows what is happening using the visual representation

![image](https://user-images.githubusercontent.com/5932954/196005765-fa9547b2-5803-4e44-a219-87012d5a111a.png)

with "P2.Part2" being the sequence of names of the entities and "packages.classes" being the sequence of arrows labels on the path. Now, if I try to do something similar with the example described above, I have the visual representation

![image](https://user-images.githubusercontent.com/5932954/196005814-6588ab17-ce15-489f-b0bf-cf74124666da.png)

where I run in a problem because the "entities" are "anonymous" (I am most certainly not using the right nomenclature, but hopefully you can understand what I mean). I tried something like `find(model, "b", "variables.vars")` but that returns None. How would I find `b` in this example?

**Reply by goto40** [reply]
I  not sure If I understood correctly, but I think you need to make sure to follow "variables" without consuming the name: `~variables`. This leads to an rrel expression `~variables.vars`.

`~variables.vars` means: 

* start at the model root (no dot at the beginning)
* follow the attribute `variables` without consuming the name (because of `~`).
* follow vars and consume the current part of the name (in case you have a dot separated name, the parts are processed one after the other): in your case look for "b" in `vars`.

The operators and how they are applied is explained in
http://textx.github.io/textX/latest/rrel/.

**Reply by christian-oreilly** [reply]
Thanks, @goto40 . My actual problem was actually quite a bit more complex than simplified exampled shared here, but your comments were sufficient to help me connect the dots. If you were curious about where this was applied, it is to update and extend this code https://github.com/borismarin/pynmodl which deals with a DSL created to code cellular membrane mechanisms for the NEURON simulator. Once I figured how to work with it, textx RREL worked great for the scoping of variables!
